File: Part2Problems/CoinChange.java
package Part2Problems;

//time complxity = o(n*m) for both coins nd lopping thoruhg target;
//space complxity is = o(n) due to dp array


public class CoinChange {
    public static int coinChange(int[] coins , int target){
        int[] dp = new int[target+1];
        //[2,5,10]
        //[20]
        dp[0] = 1;
        for(int coin : coins){
            for(int i=coin; i<= target;i++){
                dp[i] += dp[i-coin];
            }
        }
        return dp[target]; 

    }
    public static void main(String[] args){
        int[] test1 = {2,5,10};
        int target = 20;
        int tar2 = 15;
        System.out.println(coinChange(test1, target));
        System.out.println(coinChange(test1, tar2));
    }
}

File: anny_dang/homework4/coinChange.py
def coinChange(coins, k):
    """
    idea:
    - create a dp with k + 1 in length -> dp[amount] stores how many ways
    we can make that amount using coins processed so far 
    - set dp[0] = 1 because there is 1 way to make 0 (no coins used)
    - for each coin, update all amounts from coin to target
        - dp[amount] += dp[amount - coin] 

    time: O(n * k) (n: number of coins, k : target sum)
    space: O(k)
    """
    dp = [0]*(k + 1)
    dp[0] = 1

    for coin in coins:
        for amount in range(coin, k + 1):
            dp[amount] += dp[amount - coin]
    
    return dp[k]
    


coins = [2, 5, 10]

print("Coins:", coins)
print("Sum: 20 ->", coinChange(coins, 20))
print("Sum: 15 ->", coinChange(coins, 15))
